### PR TITLE
Refine the example and do some maintenance

### PR DESCRIPTION
**Pull request by pirgeo**

### CODE REVIEW

**Comment by pirgeo** [comment]
Python 3.5 is no longer maintained, should we drop it from the tests? Also, it seems like OpenTelemetry will drop Python 3.5 support: https://github.com/open-telemetry/opentelemetry-python/issues /1698

File: tox.ini [hunk 1]
```diff
@@ -1,6 +1,5 @@
 [tox]
 envlist =
-    py35
```
Comment (arminru) [inline tox.ini:3]: Please also adapt the `classifiers` and `python_requires` in `setup.cfg` accordingly.

File: README.md [hunk 2]
```diff
@@ -32,7 +34,21 @@ counter = meter.create_counter(
 counter.add(25, {"dimension-1", "value-1"})
 ```
 
-Full setup examples can be found [here](example/).
+### Example
+
+To run the [example](example/basic_example.py), clone this repository and change to the `opentelemetry-metric-python` folder, then run:
+
+```shell
+pip install .                       `# install the Dynatrace exporter locally`
```
Comment (arminru) [inline README.md:42]:
```suggestion
pip install .                       `# install the Dynatrace exporter locally`
```
Where else would you install it?

File: README.md [hunk 3]
```diff
@@ -32,7 +34,21 @@ counter = meter.create_counter(
...
+export LOGLEVEL=DEBUG               `# (optional) Set the log level, valid values are: DEBUG, INFO, WARN/WARNING, ERROR, CRITICAL/FATAL
```
Comment (arminru) [inline README.md:43]:
```suggestion
export LOGLEVEL=DEBUG               `# (optional) Set the log level to default to see more output (default is INFO)
```

File: README.md [hunk 4]
```diff
@@ -32,7 +34,21 @@ counter = meter.create_counter(
...
+python example/basic_example.py
```
Comment (arminru) [inline README.md:44]:
```suggestion
python example/basic_example.py
```

File: README.md [hunk 5]
```diff
@@ -32,7 +34,21 @@ counter = meter.create_counter(
...
+
+To run the [example](example/basic_example.py), clone this repository and change to the `opentelemetry-metric-python` folder, then run:
+
+```shell
+pip install .                       `# install the Dynatrace exporter locally`
+export LOGLEVEL=DEBUG               `# (optional) Set the log level, valid values are: DEBUG, INFO, WARN/WARNING, ERROR, CRITICAL/FATAL
+python example/basic_example.py    
+```
+
+A more complete set up routine can be found [here](example/example_usage.sh), including installing inside a virtual environment and getting required packages.
```
Comment (arminru) [inline README.md:47]:
```suggestion
A more complete setup routine can be found [here](example/example_usage.sh), including installing inside a virtual environment and getting required packages.
```

File: README.md [hunk 6]
```diff
@@ -32,7 +34,21 @@ counter = meter.create_counter(
...
+If you just want to see it in action, it should be sufficient to run [`example/example_usage.sh`](example/example_usage.sh).
```
Comment (arminru) [inline README.md:48]: I think `install_and_run.sh` or similar would be a clearer name.

File: README.md [hunk 7]
```diff
@@ -60,11 +77,18 @@ The scope required for sending metrics is the `Ingest metrics` scope in the **AP
 
 #### Metric Key Prefix
 
-The `prefix` parameter specifies an optional prefix, which is prepended to each metric key, separated by a dot (`<prefix>.<namespace>.<name>`).
+The `prefix` parameter specifies an optional prefix, which is prepended metric key, separated by a dot (`<prefix>.<namespace>.<name>`).
```
Comment (arminru) [inline README.md:80]:
```suggestion
The `prefix` parameter specifies an optional prefix, which is prepended to each metric key, separated by a dot (`<prefix>.<namespace>.<name>`).
```

File: README.md [hunk 8]
```diff
@@ -60,11 +77,18 @@ The scope required for sending metrics is the `Ingest metrics` scope in the **AP
...
+#### Default Dimensions
+
+The `default_dimensions` parameter can be used to optionally specify a list of key/value pairs, which will be added as additional dimensions to all data points.
+Dimension keys are unique, and labels on instruments will overwrite the default dimensions if key collisions appear.
 
-#### Default Labels/Dimensions
+#### Export OneAgent Metadata
 
-The `tags` parameter can be used to optionally specify a list of key/value pairs, which will be added as additional labels/dimensions to all data points.
+If running on a host with a running OneAgent, setting this option will export metadata collected by the OneAgent to the Dynatrace endpoint.
```
Comment (arminru) [inline README.md:89]: The name of the parameter is not mentioned.

File: example/basic_example.py [hunk 9]
```diff
@@ -11,23 +11,47 @@
...
+import logging
+import os
+import psutil
+
+
+def get_random_number(maximum: int, minimum: int = 0):
+    if maximum < minimum:
+        minimum, maximum = maximum, minimum
+
+    return random.randint(minimum, maximum)
```
Comment (arminru) [inline example/basic_example.py:31]: Since you're just using it four times here (and always with the correct params 😉), you could also just call `random.randint(0, x)` or `random.randrange(x+1)` directly.

File: example/basic_example.py [hunk 10]
```diff
@@ -62,18 +86,33 @@ def parse_arguments():
...
+    # setting it to "INFO" if not found.
+    # Valid levels are: DEBUG, INFO, WARN/WARNING, ERROR, CRITICAL/FATAL
+    loglevel = os.environ.get("LOGLEVEL", "INFO").upper()
+    logging.basicConfig(level=loglevel)
+    logger = logging.getLogger(script_name)
+
     if not args.endpoint:
-        print("No Dynatrace endpoint specified, exporting to default local "
-              "OneAgent ingest endpoint.")
+        logger.warning(
```
Comment (arminru) [inline example/basic_example.py:99]:
```suggestion
        logger.info(
```

File: example/basic_example.py [hunk 11]
```diff
@@ -62,18 +86,33 @@ def parse_arguments():
...
+            "No Dynatrace endpoint specified, exporting to default local "
+            "OneAgent endpoint.")
 
-    # set up opentelemetry for export:
+    # set up OpenTelemetry for export:
+    logger.info("setting up global OpenTelemetry configuration.")
```
Comment (arminru) [inline example/basic_example.py:104]: Those would rather be `debug` IMHO

File: src/dynatrace/opentelemetry/metrics/export/__init__.py [hunk 12]
```diff
@@ -36,37 +36,43 @@ class DynatraceMetricsExporter(MetricsExporter):
...
+        self._serializer = DynatraceMetricsSerializer(prefix, all_dimensions)
         self._session = requests.Session()
         self._headers = {
             "Accept": "*/*; q=0",
             "Content-Type": "text/plain; charset=utf-8",
         }
         if api_token:
-            self._headers["Authorization"] = "Api-Token " + api_token
+            if not endpoint_url:
+                self.__logger.warning("Just API token but no endpoint passed. "
```
Comment (arminru) [inline src/dynatrace/opentelemetry/metrics/export/__init__.py:71]:
```suggestion
                self.__logger.info("Just API token but no endpoint passed. "
```

File: src/dynatrace/opentelemetry/metrics/export/__init__.py [hunk 13]
```diff
@@ -100,6 +108,7 @@ def export(
                 headers=self._headers,
             ) as resp:
                 resp.raise_for_status()
+                self.__logger.debug(resp.content)
```
Comment (arminru) [inline src/dynatrace/opentelemetry/metrics/export/__init__.py:111]:
```suggestion
                self.__logger.debug("got response:" + resp.content)
```
or similar (unless the content is self-explanatory anyway)

File: src/dynatrace/opentelemetry/metrics/export/serializer.py [hunk 14]
```diff
@@ -68,12 +68,26 @@ class DynatraceMetricsSerializer:
     def __init__(
         self,
         prefix: Optional[str],
-        tags: Optional[Mapping],
+        dimensions: Optional[Mapping],
```
Comment (arminru) [inline src/dynatrace/opentelemetry/metrics/export/serializer.py:71]:
```suggestion
        default_dimensions: Optional[Mapping],
```

File: src/dynatrace/opentelemetry/metrics/export/serializer.py [hunk 15]
```diff
@@ -68,12 +68,26 @@ class DynatraceMetricsSerializer:
...
     ):
         self._prefix = prefix
-        self._tags = tags or {}
         self._is_delta_export = None
 
+        self._dimensions = self._normalize_dimensions(dimensions)
```
Comment (arminru) [inline src/dynatrace/opentelemetry/metrics/export/serializer.py:76]:
```suggestion
        self._default_dimensions = self._normalize_dimensions(default_dimensions)
```

File: src/dynatrace/opentelemetry/metrics/export/serializer.py [hunk 16]
```diff
@@ -68,12 +68,26 @@ class DynatraceMetricsSerializer:
...
+
+    @classmethod
+    def _normalize_dimensions(cls, dimensions):
+        dim_dict = {}
+        if dimensions:
+            # normalize the dimensions once, so it doesnt have to be repeated
```
Comment (arminru) [inline src/dynatrace/opentelemetry/metrics/export/serializer.py:82]:
```suggestion
            # normalize the dimensions once, so it doesn't have to be repeated
```

File: src/dynatrace/opentelemetry/metrics/export/serializer.py [hunk 17]
```diff
@@ -68,12 +68,26 @@ class DynatraceMetricsSerializer:
...
+            # for every serialisation
+            for k, v in dimensions.items():
+                key = cls._normalize_dimension_key(k)
+                if key:
+                    dim_dict[key] = cls._normalize_dimension_value(v)
```
Comment (arminru) [inline src/dynatrace/opentelemetry/metrics/export/serializer.py:87]: Could this return None?

File: src/dynatrace/opentelemetry/metrics/export/serializer.py [hunk 18]
```diff
@@ -101,12 +115,14 @@ def _write_record(
...
             return
         string_buffer.append(metric_key)
-        self._write_dimensions(string_buffer, record.labels)
-        if self._tags:
-            self._write_dimensions(string_buffer, self._tags.items())
 
-        serialize_func(string_buffer, aggregator)
+        unique_dimensions = self._make_unique_dimensions(record.labels,
+                                                         self._dimensions.
+                                                         items())
```
Comment (arminru) [inline src/dynatrace/opentelemetry/metrics/export/serializer.py:121]: This looks like 3 params at first glance, maybe indent it a bit less.

File: src/dynatrace/opentelemetry/metrics/export/serializer.py [hunk 19]
```diff
@@ -101,12 +115,14 @@ def _write_record(
...
+        DynatraceMetricsSerializer._write_dimensions(string_buffer,
```
Comment (arminru) [inline src/dynatrace/opentelemetry/metrics/export/serializer.py:122]: Why not self like below?

File: src/dynatrace/opentelemetry/metrics/export/serializer.py [hunk 20]
```diff
@@ -234,20 +247,15 @@ def __normalize_dimension_key_section(cls, section: str):
...
-                dim_value = (DynatraceMetricsSerializer.
-                             _normalize_dimension_value(value))
-
-                string_buffer.append(",")
-                string_buffer.append(dim_key)
-                string_buffer.append("=")
-                string_buffer.append(dim_value)
-
-            # else: the dimension is empty, the dimension is dropped.
+        # when the calling function uses make unique dimensions first, keys and
```
Comment (arminru) [inline src/dynatrace/opentelemetry/metrics/export/serializer.py:252]: This assumption/prerequisite should be mentioned in a docstring

File: src/dynatrace/opentelemetry/metrics/export/serializer.py [hunk 21]
```diff
@@ -271,6 +279,25 @@ def _remove_control_characters(cls, s: str) -> str:
...
     def _normalize_dimension_value(cls, value: str):
         value = value[:cls.__dv_max_length]
-        value = DynatraceMetricsSerializer._remove_control_characters(value)
+        value = cls._remove_control_characters(value)
         value = cls.__re_dv_escape_chars.sub(r"\\\g<1>", value)
         return value
+
+    @classmethod
+    def _make_unique_dimensions(cls, labels: Iterable[Tuple[str, str]],
+                                dimensions: Iterable[Tuple[str, str]]):
```
Comment (arminru) [inline src/dynatrace/opentelemetry/metrics/export/serializer.py:288]: Please document the precedence applied here.

File: README.md (hunk 8 above)
Comment (arminru) [inline README.md:89]: The precedence over labels and default dimensions needs to be documented.

File: src/dynatrace/opentelemetry/metrics/export/__init__.py (hunk 12 above)
Comment (dyladan) [reply src/dynatrace/opentelemetry/metrics/export/__init__.py:73]: We should have clear guidelines for logging levels. IMO this would be a warn

Comment (arminru) [reply src/dynatrace/opentelemetry/metrics/export/__init__.py:73]: Ah I confused the context here - passing a token while using the local endpoint indicates a misconception we should warn users about indeed, so warning is fine here. Thanks for double-checking @dyladan!

File: src/dynatrace/opentelemetry/metrics/export/serializer.py (hunk 21 above)
Comment (pirgeo) [reply src/dynatrace/opentelemetry/metrics/export/serializer.py]: should be done now.

File: src/dynatrace/opentelemetry/metrics/export/serializer.py (hunk 17 above)
Comment (pirgeo) [reply src/dynatrace/opentelemetry/metrics/export/serializer.py:91]: it can return an empty dimension key, which would also evaluate to `False` and will lead to invalid dimension keys not being added.

File: README.md [hunk 22]
```diff
@@ -62,9 +79,19 @@ The scope required for sending metrics is the `Ingest metrics` scope in the **AP
...
+Dimension keys are unique, and labels on instruments will overwrite the default dimensions if key collisions appear.
+
+#### Export OneAgent Metadata
+
+If running on a host with a running OneAgent, setting the `export_oneagent_metadata` option to `True` will export metadata collected by the OneAgent to the Dynatrace endpoint.
+If no Dynatrace API endpoint is set, the default exporter endpoint will be the OneAgent endpoint, and this option will be set automatically.
+Therefore, if no endpoint is specified, we assume a OneAgent is running and export to it, including metadata.
 
-The `tags` parameter can be used to optionally specify a list of key/value pairs, which will be added as additional labels/dimensions to all data points.
+##### Dimensions precedence
```
Comment (arminru) [inline README.md:93]:
```suggestion
##### Dimensions precedence

```

File: README.md [hunk 23]
```diff
@@ -62,9 +79,19 @@ The scope required for sending metrics is the `Ingest metrics` scope in the **AP
...
+When specifying default dimensions, labels and OneAgent metadata enrichment, the precedence of dimensions with the same key is as follows: Default dimensions are overwritten by labels passed to instruments, which in turn are overwritten by the OneAgent dimensions (even though the likeliness of a collision here is very low). If the same key is present _within_ any of the three categories, the last occurrence of each key will be used.
```
Comment (arminru) [inline README.md:94]:
```suggestion
When specifying default dimensions, labels and OneAgent metadata enrichment, the precedence of dimensions with the same key is as follows:
Default dimensions are overwritten by labels passed to instruments, which in turn are overwritten by the OneAgent dimensions
(even though the likeliness of a collision here is very low, since the OneAgent metadata only contains [Dynatrace reserved dimensions](https://www.dynatrace.com/support/help/how-to-use-dynatrace/metrics/metric-ingestion/metric-ingestion-protocol/#syntax) starting with `dt.*`).
```

File: src/dynatrace/opentelemetry/metrics/export/__init__.py [hunk 24]
```diff
@@ -54,13 +54,15 @@ def __init__(
                                "to default local OneAgent ingest endpoint.")
             self._endpoint_url = "http://localhost:14499/metrics/ingest"
 
-        all_dimensions = default_dimensions or {}
+        one_agent_dims = {}
```
Comment (arminru) [inline src/dynatrace/opentelemetry/metrics/export/__init__.py:57]: OneAgent is usually used as one word.

File: src/dynatrace/opentelemetry/metrics/export/serializer.py [hunk 25]
```diff
@@ -67,26 +67,30 @@ class DynatraceMetricsSerializer:
 
     def __init__(
         self,
-        prefix: Optional[str],
-        dimensions: Optional[Mapping],
+        prefix: Optional[str] = "",
+        default_dimensions: Optional[Mapping] = None,
+        one_agent_dimensions: Optional[Mapping] = None,
```
Comment (arminru) [inline src/dynatrace/opentelemetry/metrics/export/serializer.py:72]: Same here (oneagent).

File: src/dynatrace/opentelemetry/metrics/export/serializer.py [hunk 26]
```diff
@@ -67,26 +67,30 @@ class DynatraceMetricsSerializer:
...
-            # for every serialisation
+            # normalize the dimensions once, so it doesn't have to be repeated
+            # for every serialization
             for k, v in dimensions.items():
                 key = cls._normalize_dimension_key(k)
                 if key:
                     dim_dict[key] = cls._normalize_dimension_value(v)
 
-        return dim_dict
+        return [(k, v) for k, v in dim_dict.items()]
```
Comment (arminru) [inline src/dynatrace/opentelemetry/metrics/export/serializer.py]: Why was this changed?

File: example/basic_example.py [hunk 27]
```diff
@@ -62,18 +79,33 @@ def parse_arguments():
...
+    logger.debug("setting up global OpenTelemetry configuration.")
     metrics.set_meter_provider(MeterProvider())
     meter = metrics.get_meter(splitext(basename(__file__))[0])
 
+    logger.info("setting up Dynatrace metrics exporting interface.")
     exporter = DynatraceMetricsExporter(args.endpoint, args.token,
                                         prefix="otel.python",
-                                        export_oneagent_metadata=args.metadata_enrichment)
+                                        export_oneagent_metadata=args.
+                                        metadata_enrichment)
```
[the remark this answers is not in the document]
Comment (pirgeo) [reply example/basic_example.py:105]: moved the dot to the next line to start the line off with it, which also seems more natural to me.

File: src/dynatrace/opentelemetry/metrics/export/serializer.py (hunk 26 above)
Comment (pirgeo) [reply src/dynatrace/opentelemetry/metrics/export/serializer.py]: i think this was due to an oversight on my part, thinking that it wasn't possible to type-hint `dict[str, str]` in the method signature of `_make_unique`. I since figured out a way to do that and moved the code here back to using dictionaries.

File: README.md [hunk 28]
```diff
@@ -32,7 +34,21 @@ counter = meter.create_counter(
...
+To run the [example](example/basic_example.py), clone this repository and change to the `opentelemetry-metric-python` folder, then run:
+
+```shell
+pip install .                       # install the Dynatrace exporter
+export LOGLEVEL=DEBUG               # (optional) Set the log level to debug to see more output (default is INFO)
+python example/basic_example.py
+```
+
+A more complete setup routine can be found [here](example/install_and_run.sh), including installing inside a virtual environment and getting required packages.
+If you just want to see it in action, it should be sufficient to run [`example/install_and_run.sh`](example/install_and_run.sh).
```
Comment (arminru) [inline README.md:48]:
```suggestion
If you just want to see it in action, it should be sufficient to run [`example/install_and_run.sh`](example/install_and_run.sh) from the root folder.
```

File: README.md [hunk 29]
```diff
@@ -32,7 +34,21 @@ counter = meter.create_counter(
...
+This script will set up a virtual environment, pull in all the required packages and run the [example](example/basic_example.py).
```
Comment (arminru) [inline README.md:49]:
```suggestion
This script will install Python, set up a virtual environment, pull in all the required packages and run the [example](example/basic_example.py).
```

**Review by arminru — APPROVED** [review]
LGTM! :)
